Add: support for emscripten (play-OpenTTD-in-the-browser)

TrueBrain opened the pull request:
(depends on https://github.com/OpenTTD/bananas-server/pull/39 before "online content" will work out-of-the-box)

(#7510 was used as inspiration for this PR. Different choices were made, as this PR is more self-contained and doesn't rely on custom Dockerfiles; there is nothing wrong with the approach taken in #7510, I just prefer a different one myself)

----

## Running
To compile, first make a "build-host" folder in which you build "tools" for the host (we run this in the Docker container, to avoid library mismatches):

```
mkdir build-host
docker run -it --rm -v $(pwd):$(pwd) -u $(id -u):$(id -g) --workdir $(pwd)/build-host emscripten/emsdk cmake .. -DOPTION_DEDICATED=ON
docker run -it --rm -v $(pwd):$(pwd) -u $(id -u):$(id -g) --workdir $(pwd)/build-host emscripten/emsdk make -j5 tools
```

Next, use the following line in a new build folder to generate the WASM / JavaScript / HTML files:

```
mkdir build
docker run -it --rm -v $(pwd):$(pwd) -u $(id -u):$(id -g) --workdir $(pwd)/build emscripten/emsdk emcmake cmake .. -DHOST_BINARY_DIR=$(pwd)/build-host -DCMAKE_BUILD_TYPE=Release -DOPTION_USE_ASSERTS=OFF
docker run -it --rm -v $(pwd):$(pwd) -u $(id -u):$(id -g) --workdir $(pwd)/build emscripten/emsdk emmake make -j5
```

This produces the following files:
```
openttd.data
openttd.html
openttd.js
openttd.wasm
```
Which combined are the full game, which you can run in a browser. You do need a HTTP server to run it, even if used locally. Browsers don't allow XHR requests with `file://`. (pro-tip: `python3 -m http.server`)

## Reasoning
The idea behind adding Emscripten is that we could produce these kind of binaries on Pull Request (after a trusted contributor okays it, to prevent abuse), so new features can be tested quickly, and bug-fixes can be reviewed all from within the browser. This Pull Request puts down the groundwork for this. It is also the reason no other external libraries are used, like for example there is no support for LZMA compression (emscripten doesn't have a port available for this).

If we would add LZMA support, we could extend this as a "Play Now" features on our website (and for example, reddit could do the same, and CityMania, or-whatever community wants to drag in players quickly). It creates a smooth and simple experience for new players to jump in.

## Design choices
- [x] The SDL2 loop unrolling is strictly seen not required: emscripten can also run in a so called `asyncify` mode, where a sleep does a "context switch" to Javascript world, allowing for the canvas to be drawn. There are two significant issues with this approach:
1) with `asyncify`, SDL2 is not v-sync'd, meaning mouse movement etc feels laggy and distorted. Frames are drawn when-ever. With the current Pull Request, it is v-sync'd, that is: when ever the browser indicates an animation frame should happen, OpenTTD does a tick and draws the screen. This feels significant less laggy.
2) it has to unwind/rewind the stack, which is not always as trivial as it sounds. This can cause lockups (for example when you pause), or crashes. These problems are all solvable, but requires significant effort.
- [x] Only SDL2 is supported, although supporting SDL is also possible. But #7510 already mentions performance is worse with SDL, so no effort was taken to support it
- [x] 32bpp is used over 8bpp, as the performance is noticeable better with 32bpp.
- [x] The mouse is using relative-mode, as you can only do a "pointer lock" with emscripten, and not change the position of the cursor like we normally do.
- [x] OpenGFX is automatically loaded if it is detected that you haven't before. It is stored in the Indexed DB in your browser. This means that someone that plays more often, doesn't have to download 6MB every single time, but only once.
- [x] Emscripten has 2 bugs in their network stack that are blocking for us. Both are an upstream issue:
1) https://github.com/emscripten-core/emscripten/issues/12996
2) https://github.com/emscripten-core/emscripten/issues/12998
Also applied is a commit that adds workarounds in our codebase for those bugs; they are very ugly, and should be considered a hack. But for now, it works, and once emscripten caught up with these issues, the hacks can be removed.
- [x] Listing servers from master-server and adding servers manually is disabled. To join a random server, someone needs to run an open proxy where WebSocket is translated into UDP/TCP from/to the game server. openttd.org will not be doing this because of the security aspects of such open proxy. As such, the buttons for this are disabled by default. However, communities like CityMania, Reddit, Discord, .. can build their own web-version with hard-coded servers in them, where they host themself a WebSocket proxy for those people to connect via. This can be done in `pre.js` and is documented in comments (`openttd_server_list`).
- [x] Start server is disabled, as you cannot run a server over WebSocket. It would be possible via a proxy to do this, but it feels a bit weird to allow this.

## Known-bugs
- [x] Fast-Forward doesn't really work. That is to say, it strongly depends on your screens how fast the game can run. Because we are sync'd with Animation Frames, which browsers normally trigger on the refresh frequency of your screen, we cannot go faster than the browser allows us. If you have 144 Hz screens, it means Fast-Forward can go up to 144 fps. Using `asyncify` does not resolve this issue, but still restricts us to ~120 fps.

## Example of preview
https://github.com/TrueBrain/OpenTTD/pull/19

TrueBrain committed:
Add: support for emscripten (play-OpenTTD-in-the-browser)

Emscripten compiles to WASM, which can be loaded via
HTML / JavaScript. This allows you to play OpenTTD inside a
browser.

Co-authored-by: milek7 <me@milek7.pl>

James103 commented:
Would it be possible to add support for reading and writing data to/from the `C:\Users\[user]\Documents\OpenTTD` folder in addition to the browser's local storage? That way, you could have access to saved games and NewGRFs from your local installation (if you have any) rather than just having to use the browser's local storage.

TrueBrain commented:
> 
> 
> Would it be possible to add support for reading and writing data to/from the `C:\Users\[user]\Documents\OpenTTD` folder in addition to the browser's local storage? That way, you could have access to saved games and NewGRFs from your local installation (if you have any) rather than just having to use the browser's local storage.

Not as far as I know. Browsers sandbox everything, and I would assume security wouldn't let access to the users disks. This is also one of the reasons I would love to add Cloud Saves, as that would solve this too :)

What is possible, how-ever, is to allow uploading of a savegame, and I am sure it is also possible to make a "Download savegame" button which uses the normal browser flow to download a file. But that is not really user-friendly :) (and currently also very low on the list to check out)

James103 commented:
Note that storing too much data in local storage may result in a warning or error from the browser, and the excess data may not be written correctly. The limit in this case can be as low as 5 MB (or even lower).

TrueBrain commented:
> 
> 
> Note that storing too much data in local storage may result in a warning or error from the browser, and the excess data may not be written correctly. The limit in this case can be as low as 5 MB (or even lower).

https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Browser_storage_limits_and_eviction_criteria

`There's also another limit called group limit — this is defined as 20% of the global limit, but it has a minimum of 10 MB and a maximum of 2 GB.`

But local storage is an LRU, and files can be evicted for many reasons. But for now, it will do fine. But here too, Cloud Saves are kinda mandatory to really make this shine :)

LordAro commented on src/settings.cpp:
```diff
@@ -1913,6 +1913,12 @@ CommandCost CmdChangeSetting(TileIndex tile, DoCommandFlag flags, uint32 p1, uin
 
 		SetWindowClassesDirty(WC_GAME_OPTIONS);
+
+#ifdef __EMSCRIPTEN__
```
Why not just do this anyway? It's not like saving to file is particularly expensive...

TrueBrain replied:
I was just ranting about that on IRC :P It is a change in functionality for all OSes ... I am all in favour, but .. I try to minimize the impact with this PR as much as possible :D I will make a PR to apply this for all OSes, and we can see what we think. I can also hide it behind a command line argument ... dealers choice!

TrueBrain committed:
Add: [Emscripten] use "relative mouse mode" with SDL2

This mode doesn't wrap the mouse constantly, but requests SDL
to lock the mouse pointer. This is needed, as with Emscripten
you are not allowed to change the mouse poisition (only to lock
it into place).

TrueBrain committed:
Codechange: unroll the SDL2 main loop

This commit prepares for the next commit, as Emscripten needs to
have a way to trigger a single iteration of the main loop. To
keep the real changes more clear, this commit only unrolls the
loop, and makes no changes to the logic itself.

TrueBrain committed:
Add: [Actions] preview workflow, where PRs can be reviewed in the browser

When a developer attaches the "preview" label, a build is created
and published on https://preview.openttd.org/. After that, new
pushes to the PR are automatically build (as long as the "preview"
label exists).

If a non-developer attaches the "preview" label, it will be
removed.

TrueBrain committed:
Fix: workarounds for two emscripten bugs in the network stack

TrueBrain commented on src/network/core/config.h:
```diff
@@ -24,7 +24,11 @@ static const char * const NETWORK_CONTENT_MIRROR_URL            = "/bananas";
 
 static const uint16 NETWORK_MASTER_SERVER_PORT    = 3978;         ///< The default port of the master server (UDP)
+#ifdef __EMSCRIPTEN__
+static const uint16 NETWORK_CONTENT_SERVER_PORT   =  443;         ///< The default port of the WebSocket server.
```
I am really unsure about this. It is correct, but the location feels kinda meh ...

It could also be moved to `pre.js`, and as such, out of the OpenTTD's codebase (and into the `os/emscripten` folder).

I would love opinions here!

(background: emscripten uses WebSocket to communicate, so we setup a WebSocket on `https://content.openttd.org`, so this client can connect there, and talk the normal protocol to talk to the content service; but this does require another port, and using the default https is the easiest).

LordAro replied:
Feel like pre.js might be nicer. At least it needs the background comment above in the code

TrueBrain replied:
Turns out it cannot go in `pre.js`, because Emscripten is a bit weird. There is a pending PR to fix that, so I will roll this out if that ever lands :) For now, added the background story, good call!

TrueBrain commented on src/network/core/os_abstraction.h:
```diff
@@ -83,6 +83,15 @@ typedef unsigned long in_addr_t;
+#		define AI_ADDRCONFIG 0
+/* Emscripten says it supports FD_SETSIZE fds, but it really only supports 64. */
+#		undef FD_SETSIZE
+#		define FD_SETSIZE 64
```
In 2013 somewhere they said they wanted to fix this .. they never did. So I wouldn't get my hopes up this will ever be fixed upstream.

LordAro replied:
Again, would be good to have a link to the (closed) issue for reference

TrueBrain committed:
Fix: [Git] ignore any build folder you might have

Instructions now suggest using build-host, etc. It is easier to
just ignore all build* folders.

LordAro commented on .github/workflows/preview_build.yml:
```diff
@@ -0,0 +1,132 @@
+        emcmake cmake .. \
+          -DHOST_BINARY_DIR=../build-host \
+          -DCMAKE_BUILD_TYPE=RelWithDebInfo \
+          -DOPTION_USE_ASSERTS=OFF \
```
The documentation isn't particularly clear, but I think asserts are already disabled in RelWithDebInfo BUILD_TYPES (certainly is with Release, why not use that?)

LordAro commented on .github/workflows/preview_build.yml:
```diff
@@ -0,0 +1,132 @@
+
+    - name: Publish preview
+      run: |
+        pip3 install setuptools
+        pip3 install awscli
```
If these 2 can't be squashed into a single command, would be good to have a comment why

TrueBrain replied:
good point. And owh boy, this was a total shitshow ... the pip installed in that Docker image is .. "special"

LordAro commented on .github/workflows/preview_build.yml:
```diff
@@ -0,0 +1,132 @@
+        pip3 install setuptools
+        pip3 install awscli
+
+        ~/.local/bin/aws s3 cp --only-show-errors build/openttd.data s3://${{ secrets.PREVIEW_S3_BUCKET }}/${{ github.event.client_payload.folder }}/
```
Might be nicer to just add ~/.local/bin to PATH?

TrueBrain replied:
I more often do that wrong than doing it this way, honestly :) And there might be other "aws" installations around. I really want the one that pip installed :P So I rather not :)

LordAro commented on .github/workflows/preview_build.yml:
```diff
@@ -0,0 +1,132 @@
+        AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+        AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+
+    - name: Update deployment status to success
```
I don't see why this manual "setting deployment status" is necessary - shouldn't it be the default based on the previous action. Can you explain?

TrueBrain replied:
How deployments on GitHub work:

- You create a deployment -> this creates a "deployment" event
- You start a deployment (set status to pending)
- When it is done, you set the status to either success or failure

This step sets it to success, as we are done now, and the URL is available.

Does that answer the question a bit? :)

LordAro commented on CMakeLists.txt:
```diff
@@ -15,6 +15,10 @@ if(NOT CMAKE_BUILD_TYPE)
     set(CMAKE_BUILD_TYPE Debug)
 endif()
 
+if (EMSCRIPTEN)
+    set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/os/emscripten/cmake")
+endif()
```
Could consider generalising this to all OSes, and just do a check if the folder exists

TrueBrain replied:
If we get more than one, yes. But given we only have one, I rather not premature generalise :)

LordAro commented on src/network/network_gui.cpp:
```diff
@@ -475,6 +478,14 @@ class NetworkGameWindow : public Window {
+		/* As the master-server doesn't support "websocket" servers yet, we
+		 * let "os/emscripten/pre.js" hardcode a list of servers people can
+		 * join. This means the serverlist is curated for now, but it is the
+		 * best we can offer. */
```
bit ick. Might as well disable the network window entirely, especially at the moment as there are none available at all

TrueBrain replied:
I am sure some servers are going to run this to join their multiplayer, so I didn't want to disable it. But it is still a bit rough this way, I agree. Disabling it completely, I think is a bad call, and I don't really know the serverlist till it is opened.

Do you mind if we leave it like this for now, considering it more a WIP?

LordAro commented on src/saveload/saveload.cpp:
```diff
@@ -45,6 +45,9 @@
 #include <atomic>
 #include <string>
+#ifdef __EMSCRIPTEN__
+# include <emscripten.h>
```
This (and all the others) should be indented with tabs (as per stdafx.h)

LordAro commented on src/video/sdl2_v.cpp:
```diff
@@ -664,19 +701,104 @@ void VideoDriver_SDL::Stop()
+	if (_exit_game) {
+#ifdef __EMSCRIPTEN__
+		emscripten_cancel_main_loop();
+		MainLoopCleanup();
```
Previously this was called when the main loop was broken out of here, now it is not

TrueBrain replied:
This is when the mainloop is broken out of. Emscripten unrolls the main loop, and as such, it no longer exists. When `_exit_game` goes true, it normally breaks out of the mainloop, calling this function. And now we do it here :) Will leave a comment!

LordAro reviewed (changes requested):
Told you i'd find things :)

LordAro commented on os/emscripten/cmake/FindLibLZMA.cmake:
```diff
@@ -0,0 +1,19 @@
+check_cxx_source_compiles("
+    int main() { return 0; }"
+    LIBLZMA_FOUND
+)
```
This doesn't appear to actually test lzma? Shouldn't it at least be trying to include the header? Alternatively, is it necessary at all? None of the other emscripten Find*.cmake files have compile tests

TrueBrain replied:
> None of the other emscripten Find*.cmake files have compile tests

FYI, we do this in some other `Find` scripts too, like the `Xaudio2` :) So I did not invent something new :D (I might have copied it from there :P)

Nevertheless, it is a bit uncommon, and for this you need to know a bit about how emscripten works.

So, emscripten has a very unusual way of  installing dependencies, or "ports" in their lingo. All those `-s USE_NNN` tell the wrapping scripts to ensure that dependency exists on the system. So if we do `-sUSE_LIBLZMA=1`, and compile anything with that, no matter what, emscripten is going to download and install the port of LibLZMA, fully transparent and automated.

Honestly, this is one of the more weird things emscripten does. It is also the reason you don't see us installing any dependencies, and why we need to have a patch to install a port. Rather I would have seen a command that installs the dependency of course, like all other systems, but alas, here we are :) Welcome to the world of emscripten :D

As far as this piece of code, I could add an `#include` to make it look a bit less weird. It is still weird :) I didn't want to add comments about how emscripten works, as .. that is like documenting how GCC works. But I think with adding an `#include` everyone is fine with it :)

LordAro reviewed (approved):
Oh my.

embeddedt commented:
@TrueBrain Question: is it necessary to disable the "Add server" button on Emscripten? I understand that it can obviously only work with WebSocket-based servers, but what if the URL for my WebSocket server is unpredictable? It would be great to have a way of changing it without needing to rebuild the client every time.

TrueBrain commented:
The main issue is the common user. If we leave it enabled, they will be confused thinking they can join normal servers with it. They cannot. Ideally we would have WebSocket enabled for all servers. Or that the masterserver returns a list of servers that do have a WebSocket proxy. But we are far away from that :D

I can see two ways for you to solve your problem in the here and now:

1) use a dynamic DNS provider, making your address predictable.
2) from what I read, you are already building your own client: enable it again. Just make sure your users understand they can only use it to join your server; not some random other one ;)

This is also one of the reasons we still don't have a "Play Now" button on the frontpage. There is more work to do for Emscripten before it can go mainstream. Currently it really is to provide previews in Pull Requests :) PRs to improve that situation are very welcome ;)

embeddedt commented:
Makes sense. Thanks for the quick reply!

For now I will just re-enable the button. However, I would like to help work on a better solution if possible.
